api/v1alpha1/openstackfloatingippool_types.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	// We use v1alpha7 here rather than anything newer because as of writing
+	// it is the newest API version we should no longer be making breaking
+	// changes to. If we bump this we need to look carefully for resulting
+	// CRD changes in v1alpha1 to ensure they are compatible.
+	infrav1alpha7 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha7"
+)
+
+const (
+	// OpenStackFloatingIPPoolFinalizer allows ReconcileOpenStackFloatingIPPool to clean up resources associated with OpenStackFloatingIPPool before
+	// removing it from the apiserver.
+	OpenStackFloatingIPPoolFinalizer = "openstackfloatingippool.infrastructure.cluster.x-k8s.io"
+
+	OpenStackFloatingIPPoolNameIndex = "spec.poolRef.name"
+
+	// OpenStackFloatingIPPoolIP.
+	DeleteFloatingIPFinalizer = "openstackfloatingippool.infrastructure.cluster.x-k8s.io/delete-floating-ip"
+)
+
+// ReclaimPolicy is a string type alias to represent reclaim policies for floating ips.
+type ReclaimPolicy string
+
+const (
+	// ReclaimDelete is the reclaim policy for floating ips.
+	ReclaimDelete ReclaimPolicy = "Delete"
+	// ReclaimRetain is the reclaim policy for floating ips.
+	ReclaimRetain ReclaimPolicy = "Retain"
+)
+
+// OpenStackFloatingIPPoolSpec defines the desired state of OpenStackFloatingIPPool.
+type OpenStackFloatingIPPoolSpec struct {
+	PreAllocatedFloatingIPs []string `json:"preAllocatedFloatingIPs,omitempty"`
```

Comment:
Little formatting fixes, plus I think this should have `+optional`:

```
	// PreAllocatedFloatingIPs is a list of preallocated floating IPs precreated in OpenStack that should be used by this pool.
	// These are used before allocating new ones and are not deleted from OpenStack when the pool is deleted.
```

Reply:
@dulek FYI I've been recommending against `+optional`s because they're already the default. I don't have a strong position on it, though. Is there specific guidance somewhere I'm ignoring?

Reply:
Okay, so just the formatting fixes! :)